src/main/java/com/ldtteam/structurize/placement/StructurePlacer.java

```diff
@@ -0,0 +1,358 @@
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.World;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Supplier;
+
+/**
+ * result = StructurePlacer.execute(this)
+ *
+ * if result = missingItems
+ * return MISSING_ITE
+ * else if result = block break
+ * return BREAK_BLOCK
+ * else if result = finished
+ * return nextState
+ * else
+ * return getState()
+ *
+ * on success we have to write the increased progress pos to NBT.
+ */
+
+/**
+ * Interface for using the structure codebase.
+ */
+public class StructurePlacer
+{
+    /**
+     * The structure iterator.
+     */
+    protected final BlueprintIterator iterator;
+
+    /**
+     * The handler.
+     */
+    protected final IStructureHandler handler;
+
+    /**
+     * Create a new structure placer.
+     *
+     * @param handler the structure handler.
+     */
+    public StructurePlacer(final IStructureHandler handler)
+    {
+        this.iterator = new BlueprintIterator(handler);
+        this.handler = handler;
+    }
+
+    /**
+     * Execute structure placement.
+     * @param world the world.
+     * @param storage the change storage.
+     * @param inputPos the pos to start from.
+     * @param operation the operation to execute.
+     * @param iterateFunction the function to iterate.
+     * @param includeEntities if entities should be included.
+     * @return the result.
+     */
+    public StructurePhasePlacementResult executeStructureStep(
+      final World world,
+      final ChangeStorage storage,
+      final BlockPos inputPos,
+      final Operation operation,
+      final Supplier<BlueprintIterator.Result> iterateFunction,
+      final boolean includeEntities)
+    {
+        if (includeEntities)
+        {
+            iterator.includeEntities();
+        }
+
+        iterator.setProgressPos(new BlockPos(inputPos.getX(), inputPos.getY(), inputPos.getZ()));
+        int count = 0;
+
+        do
+        {
+            @NotNull final BlockPos localPos = iterator.getProgressPos();
+            if (count >= handler.getStepsPerCall())
+            {
+                return new StructurePhasePlacementResult(localPos, new BlockPlacementResult(localPos, BlockPlacementResult.Result.LIMIT_REACHED));
+            }
+
+            final BlockState localState = handler.getBluePrint().getBlockState(localPos);
+            if (localState == null)
+            {
+                continue;
+            }
+
+            final BlockPos worldPos = handler.getWorldPos().add(localPos).subtract(handler.getBluePrint().getPrimaryBlockOffset());
+            if (storage != null)
+            {
+                storage.addPositionStorage(worldPos, world);
+            }
+
+            final BlockPlacementResult result;
+            switch (operation)
+            {
+                case BLOCK_REMOVAL:
```

Comment:
hm wondering about having this whole thing within the structure handler, so it determines which action it wants for given blueprint data, pos

Reply:
Would take away a lot of the power this class has. I couldn't think of any better way to handle this tbf without having a ton of duplicate code.